src/python/pants/backend/scala/dependency_inference/scala_parser.py

```diff
@@ -0,0 +1,284 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+import logging
+import os
+import pkgutil
+from dataclasses import dataclass
+
+from pants.core.util_rules.source_files import SourceFiles
+from pants.engine.fs import (
+    AddPrefix,
+    CreateDigest,
+    Digest,
+    Directory,
+    FileContent,
+    MergeDigests,
+    RemovePrefix,
+)
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.process import BashBinary, FallibleProcessResult, Process, ProcessResult
+from pants.engine.rules import collect_rules, rule
+from pants.jvm.compile import CompiledClassfiles
+from pants.jvm.jdk_rules import JdkSetup
+from pants.jvm.resolve.coursier_fetch import (
+    ArtifactRequirements,
+    Coordinate,
+    MaterializedClasspath,
+    MaterializedClasspathRequest,
+)
+from pants.util.logging import LogLevel
+
+logger = logging.getLogger(__name__)
+
+PARSER_SCALA_VERSION = "2.13.7"
+SCALAMETA_VERSION = "4.4.30"
+CIRCE_VERSION = "0.14.1"
+
+PARSER_SCALA_VERSION_MAJOR_MINOR = ".".join(PARSER_SCALA_VERSION.split(".")[0:2])
+
+SCALAMETA_DEPENDENCIES = [
+    Coordinate.from_coord_str(s)
+    for s in [
+        "org.scalameta:scalameta_2.13:4.4.30",
+        "org.scala-lang:scala-library:2.13.7",
+        "com.thesamet.scalapb:scalapb-runtime_2.13:0.11.4",
+        "org.scalameta:parsers_2.13:4.4.30",
+        "org.scala-lang:scala-compiler:2.13.7",
+        "net.java.dev.jna:jna:5.8.0",
+        "org.scalameta:trees_2.13:4.4.30",
+        "org.scalameta:common_2.13:4.4.30",
+        "com.lihaoyi:sourcecode_2.13:0.2.7",
+        "org.jline:jline:3.20.0",
+        "org.scalameta:fastparse-v2_2.13:2.3.1",
+        "org.scala-lang.modules:scala-collection-compat_2.13:2.4.4",
+        "org.scala-lang:scalap:2.13.7",
+        "org.scala-lang:scala-reflect:2.13.7",
+        "com.google.protobuf:protobuf-java:3.15.8",
+        "com.thesamet.scalapb:lenses_2.13:0.11.4",
+        "com.lihaoyi:geny_2.13:0.6.5",
+    ]
+]
+
+
+CIRCE_DEPENDENCIES = [
+    Coordinate.from_coord_str(s)
+    for s in [
+        "io.circe:circe-generic_2.13:0.14.1",
+        "org.typelevel:simulacrum-scalafix-annotations_2.13:0.5.4",
+        "org.typelevel:cats-core_2.13:2.6.1",
+        "org.scala-lang:scala-library:2.13.6",
+        "io.circe:circe-numbers_2.13:0.14.1",
+        "com.chuusai:shapeless_2.13:2.3.7",
+        "io.circe:circe-core_2.13:0.14.1",
+        "org.typelevel:cats-kernel_2.13:2.6.1",
+    ]
+]
```

Comment:
To generate a lockfile for this, you can run the resolve and embed the result: see https://github.com/pantsbuild/pants/blob/3874cd6f683692ac109f41a61eb26265d67f3696/src/python/pants/backend/java/test/junit_test.py#L82-L123

Reply:
I'm going to leave this as-is since I plan on implementing proper tool lockfile support for JVM. Given we have junit, this, and tools like scalafmt on the horizon, it makes sense to implement that now.